lib/twom.h

```diff
@@ -0,0 +1,160 @@
+ *   copyright law, including all related and neighboring rights, to the extent
+ *   allowed by law.
+ *
+ *   You can copy, modify, distribute and perform the work, even for commercial
+ *   purposes, all without asking permission.
+ */
+
+#ifndef INCLUDED_TWOM_H
+#define INCLUDED_TWOM_H
+
+#include <stdbool.h>
+#include <stdlib.h>
+#include <stdint.h>
+
+struct twom_db;
+struct twom_txn;
+struct twom_cursor;
+
+enum twom_ret {
+    TWOM_OK = 0,
+    TWOM_DONE = 1,
+    TWOM_EXISTS = -1,
+    TWOM_IOERROR = -2,
+    TWOM_INTERNAL = -3,
+    TWOM_LOCKED = -4,
+    TWOM_NOTFOUND = -5,
+    TWOM_READONLY = -6,
+};
+
+// we don't reuse flags for different operations (e.g. open, fetch, foreach), as
+// there's 32 bits of space available - though not all flags have meaning in all contexts.
+enum twom_flagspec {
+    TWOM_CREATE          = 1<<0,    /* Create the database if not existant */
+    TWOM_SHARED          = 1<<1,    /* Open in shared lock mode */
+    TWOM_NOCSUM          = 1<<2,    /* Don't check checksums on read */
+    TWOM_NOSYNC          = 1<<3,    /* Don't msync/fsync on write */
+    TWOM_NONBLOCKING     = 1<<4,    /* When taking a lock, return immediately if the file is already locked */
+
+    TWOM_ALWAYSYIELD     = 1<<9,    /* Yield foreach before every callback */
+    TWOM_NOYIELD         = 1<<10,   /* Never yield a read transaction lock */
+    TWOM_IFNOTEXIST      = 1<<11,   /* Only store if the record doesn't exist */
+    TWOM_IFEXIST         = 1<<12,   /* Only store if the record already exists (e.g. when deleting) */
+    TWOM_FETCHNEXT       = 1<<13,   /* Return the record AFTER the given key */
+    TWOM_SKIPROOT        = 1<<14,   /* For foreach or cursor, skip the first record if it matches exactly */
+    TWOM_MVCC            = 1<<15,   /* For cursor or transaction, operate in serializable isolation (MVCC) mode */
+
+    TWOM_CSUM_NULL       = 1<<27,   /* use the NULL checksum when creating or repacking database */
+    TWOM_CSUM_XXH64      = 1<<28,   /* use the XXH64 checksum algorithm when creating or repacking */
+    TWOM_CSUM_EXTERNAL   = 1<<29,   /* use an external checksum algorithm (must be passed in init) */
+    TWOM_COMPAR_EXTERNAL = 1<<30    /* use an external comparison function (must be passed in init) */
+};
+
+typedef int twom_cb(void *rock,
+                    const char *key, size_t keylen,
+                    const char *data, size_t datalen);
+typedef int twom_compar(const char *s1, size_t l1, const char *s2, size_t l2);
+typedef uint32_t twom_csum(const char *s, size_t l);
+
+struct twom_open_data {
+    uint32_t flags;
+    int (*compar)(const char *s1, size_t l1, const char *s2, size_t l2);
+    uint32_t (*csum)(const char *s, size_t l);
+    void (*error)(const char *msg, const char *fmt, ...);
+};
+
+#define TWOM_OPEN_DATA_INITIALIZER { 0, NULL, NULL, NULL }
+
+// database operations
+extern int twom_db_open(const char *fname, struct twom_open_data *setup,
```

Comment:
What is the difference between with this `extern` and without it?

Reply:
https://stackoverflow.com/questions/496448/how-to-correctly-use-the-extern-keyword-in-c says "it's the default, you don't need to be explicit" - fair enough.  So "nothing".

Reply:
I removed them all